src/boat_simulator/boat_simulator/nodes/low_level_control/controller.py

```diff
@@ -0,0 +1,259 @@
+    """
+
+    def __init__(
+        self,
+        current_heading: Scalar,
+        desired_heading: Scalar,
+        current_control_ang: Scalar,
+        time_step: Scalar,
+        kp: Scalar,
+        cp: Scalar,
+        control_speed: Scalar,
+        max_angle_range=None,
+    ):
+        """Initializes the class attributes.
+
+        Args:
+            `current_heading` (Scalar): Current boat heading direction in degrees,
+             0 degrees (North) at the positive y axis and increasing clockwise.
+            `desired_heading` (Scalar): Target boating heading direction in degrees,
+             0 degrees (North) at the positive y axis and increasing clockwise.
+            `current_control_ang` (Scalar): Current control mechanism angle in degrees.
+            `time_step` (Scalar): Time per iteration given in seconds.
+            `kp` (Scalar): Proportional constant when calculating error.
+            `cp` (Scalar): Tuning parameter for control action.
+            `control_speed` (Scalar): Speed of controller change in degrees per second.
+            'running_error' (Scalar): Error between current and target control angle.
+
+        """
+
+        super().__init__(
+            current_control_ang,
+            time_step,
+            control_speed,
+            max_angle_range,
+        )
+
+        self.max_angle_range = RUDDER_MAX_ANGLE_RANGE
+        self.running_error = 0.0
+        self.current_heading = bound_to_180(current_heading)  # bound (-180, 180] in degrees
+        self.desired_heading = bound_to_180(desired_heading)  # bound (-180, 180] in degrees
+        self.kp = kp
+        self.cp = cp
+        self.setpoint = 0.0  # current setpoint angle in degrees
+        self.running_error = 0.0
+
+    def compute_error(self) -> Scalar:  # angle passed in as radians
+        """Computes the error between desired and current heading
+        implementation taken from: https://stackoverflow.com/a/2007279
+        Angles are bound with the convention (-180, 180]
+
+        Returns:
+            Scalar: The error between the given headings in radians, 0 radians at positive
+            y axis and increasing clockwise
+        """
+        desired_rad, current_rad = np.deg2rad(self.desired_heading), np.deg2rad(
+            self.current_heading
+        )
+
+        error = atan2(sin(desired_rad - current_rad), cos(desired_rad - current_rad))
+
+        return error  # in radians
+
+    def compute_setpoint(self) -> Scalar:
+        """Computes the corresponding control error angle between current control angle and
+        target control angle. Uses Raye's implementation from:
+            https://github.com/UBCSailbot/raye-boat-controller/blob/master/python/tack_controller.py
+
+        Returns:
+            Scalar: Corresponding error angle between the
+            current and target control angle in degrees"""
+
+        heading_error = self.compute_error()  # heading_error in radians
```

Comment:
Could change variable to `heading_error_rad` rather than having this comment here